src/main/java/org/jahia/modules/sitemap/filter/SitemapCacheFilter.java

```diff
@@ -59,126 +54,85 @@
-     * Uses parent session to save
-     * https://stackoverflow.com/questions/4685959/get-file-out-of-jcr-file-node
-     */
-    private JCRNodeWrapper refreshSitemapCache(JCRNodeWrapper sitemapNode, InputStream data) throws RepositoryException {
-        String cacheName = getCacheName(sitemapNode);
-        String nodePath = sitemapNode.getPath();
-
-        // use system session to be able to save file cache in live workspace
-        return JCRTemplate.getInstance().doExecuteWithSystemSessionAsUser(null,
-                Constants.LIVE_WORKSPACE, null, new JCRCallback<JCRNodeWrapper>() {
-            @Override public JCRNodeWrapper doInJCR(JCRSessionWrapper session) throws RepositoryException {
-                JCRNodeWrapper node = session.getNode(nodePath);
-                node.uploadFile(cacheName, data, "application/xml");
-                session.save();
-                return node;
-            }
-        });
-    }
+        if (needsCaching(resource)) return previousOut;
 
+        // if we are here, the cache does not exist or has expired.
 
-    /** Apply file caching only for default template */
-    public boolean needsCaching(Resource resource) {
-        String templateName = resource.getTemplate();
-        return "lang".equalsIgnoreCase(templateName);
-    }
+        String targetSitemapCacheKey = PREFIX_CACHE_NAME + resource.getNode().getLanguage();
 
-    public boolean isValidCache(JCRNodeWrapper cacheNode) {
-        long expiration = 4 * 60 * 60 * 1000L; // Default to 4 hours
-        if (cacheNode == null) return false;
-        try {
-            JCRNodeWrapper siteNode = cacheNode.getResolveSite();
-            if (siteNode != null) {
-                JCRPropertyWrapper cacheDurationProperty = siteNode.getProperty(SitemapConstant.SITEMAP_CACHE_DURATION);
-                // String propertyValue = ConversionUtils.getValueFromJCRProperty(cacheDurationProperty);
-                // expiration = ConversionUtils.toMilliSecondsLong(propertyValue, expiration);
-            }
-        } catch (RepositoryException e) {
-            logger.error("Unable to retrieve node information.");
-            return false;
+        if (sitemapCache.get(targetSitemapCacheKey) != null) {
+            sitemapCache.get(targetSitemapCacheKey);
+            return sitemapCache.get(targetSitemapCacheKey).getObjectValue().toString();
         }
-        return !isExpired(cacheNode, expiration);
-    }
 
-    public JCRNodeWrapper getCacheNode(JCRNodeWrapper sitemapNode) throws RepositoryException {
-        if (sitemapNode == null) return null;
-        String cacheName = getCacheName(sitemapNode);
-        return (sitemapNode.hasNode(cacheName)) ? sitemapNode.getNode(cacheName) : null;
-    }
+        // we get the desired cache expiration time
+        int expiredAt = sitemapService.getSitemapCacheExpirationInSeconds(resource.getNode().getResolveSite().getPropertyAsString(SitemapConstant.SITEMAP_CACHE_DURATION));
```

Comment:
All that logic should be moved to the service

Reply:
Done